### PR TITLE
Implement direct writes to S3 from browser

### PR DESCRIPTION
**Pull request by ianopolous**

### CODE REVIEW

File: src/peergos/server/storage/S3BlockStorage.java [hunk 1]
```diff
@@ -51,38 +50,83 @@
...
+                blockProps.add(new Pair<>(cid, blockSizes.get(i)));
+            }
+            List<PresignedUrl> res = new ArrayList<>();
+            for (Pair<Multihash, Integer> props : blockProps) {
+                if (props.left.type != Multihash.Type.sha2_256)
+                    throw new IllegalStateException("Can only pre-auth writes of sha256 hashed blocks!");
+                transactions.addBlock(props.left, tid, owner);
+                String s3Key = hashToKey(props.left);
+                String contentSha256 = ArrayOps.bytesToHex(props.left.getHash());
+                String host = bucket + "." + region + ".linodeobjects.com";
```
Comment (kevodwyer) [inline src/peergos/server/storage/S3BlockStorage.java:120]: why the hard-code?

File: src/peergos/server/storage/S3Request.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    public final String verb, host;
+    public final String key;
+    public final String contentSha256;
+    public final boolean allowPublicReads;
+    public final String accessKeyId;
+    public final String region;
+    public final Map<String, String> extraHeaders;
+    public final Instant date;
+
+    public S3Request(String verb,
```
Comment (kevodwyer) [inline src/peergos/server/storage/S3Request.java:173]: Feels weird the member vars and constructor at the end of file

File: src/peergos/server/tests/S3V4SignatureTests.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        String canonicalRequest = policy.toCanonicalRequest();
+        Assert.assertTrue(canonicalRequest.equals("PUT\n" +
+                "/AFKREIGBKDTWGUVD3UTHMNCKKLMIL6YE5MI4XSEYCZOWP7WTOUJ6XNMASU\n" +
+                "\n" +
+                "amz-sdk-invocation-id:e84d52c4-953e-e1d2-d442-f05ca7067524\n" +
+                "amz-sdk-retry:0/0/500\n" +
+                "content-length:10\n" +
+                "content-type:application/octet-stream\n" +
+                "host:peergos-test.us-east-1.linodeobjects.com\n" +
+                "user-agent:aws-sdk-java/1.11.705 Linux/5.3.0-46-generic OpenJDK_64-Bit_Server_VM/11.0.7+10-post-Ubuntu-2ubuntu218.04 java/11.0.7 vendor/Ubuntu\n" +
```
Comment (kevodwyer) [inline src/peergos/server/tests/S3V4SignatureTests.java:80]: How stable is this test going to be?

File: src/peergos/server/storage/S3Request.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+        byte[] dateKey = hmacSha256("AWS4" + s3SecretKey, shortDate.getBytes());
+        byte[] dateRegionKey = hmacSha256(dateKey, policy.region.getBytes());
+        byte[] dateRegionServiceKey = hmacSha256(dateRegionKey, "s3".getBytes());
+        byte[] signingKey = hmacSha256(dateRegionServiceKey, "aws4_request".getBytes());
+
+        return ArrayOps.bytesToHex(hmacSha256(signingKey, stringToSign.getBytes()));
+    }
+
+    private static byte[] sha256(byte[] input) {
```
Comment (kevodwyer) [inline src/peergos/server/storage/S3Request.java:73]: Make public? As I can see the method has been copied elsewhere.

File: src/peergos/server/tests/S3V4SignatureTests.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+        PresignedUrl url = S3Request.preSignUrl(s3Key, payload.length, contentSha256, false, timestamp,
+                "PUT", host, extraHeaders, region, accessKey, secretKey);
+        Assert.assertTrue(("AWS4-HMAC-SHA256 Credential=AKIAIOSFODNN7EXAMPLE/20200425/us-east-1/s3/aws4_request," +
+                "SignedHeaders=amz-sdk-invocation-id;amz-sdk-retry;content-length;content-type;host;user-agent;x-amz-content-sha256;x-amz-date," +
+                "Signature=5cc3daea623ac6d43b482209892cc6eb95e46b068e232eabd85343caf79bb17e")
+                .equals(url.fields.get("Authorization")));
+    }
+
+    private static byte[] sha256(byte[] input) {
```
Comment (kevodwyer) [inline src/peergos/server/tests/S3V4SignatureTests.java:104]: Duplicate method

File: src/peergos/server/storage/S3Request.java (hunk 2 above)
Comment (ianopolous) [reply src/peergos/server/storage/S3Request.java]: Agreed, that was an accident. Fixed.

File: src/peergos/server/tests/S3V4SignatureTests.java (hunk 3 above)
Comment (ianopolous) [reply src/peergos/server/tests/S3V4SignatureTests.java:80]: It's 100% deterministic.

File: src/peergos/server/storage/S3Request.java (hunk 4 above)
Comment (ianopolous) [reply src/peergos/server/storage/S3Request.java]: I've removed it and used the existing method.

File: src/peergos/server/tests/S3V4SignatureTests.java (hunk 5 above)
Comment (ianopolous) [reply src/peergos/server/tests/S3V4SignatureTests.java]: Removed

File: src/peergos/server/storage/S3BlockStorage.java (hunk 1 above)
Comment (ianopolous) [reply src/peergos/server/storage/S3BlockStorage.java]: Good spot. Fixed

**Pull request merged**